photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,253 @@
+            It also evaluates chromatic components of the image. Default: True
+        beta1: Algorithm parameter. Weight of chromatic component in the loss.
+        beta2: Algorithm parameter. Small constant, see [1].
+        beta3: Algorithm parameter. Small constant, see [1].
+        
+    Reference:
+        [1] GRADIENT MAGNITUDE SIMILARITY DEVIATION ON MULTIPLE SCALES (2017)
+            http://www.cse.ust.hk/~psander/docs/gradsim.pdf
+    """
+
+    def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.,
+                 scale_weights: Optional[Union[Tuple[float], List[float]]] = None,
+                 chromatic: bool = False, beta1: float = 0.01, beta2: float = 0.32,
+                 beta3: float = 15.) -> None:
+        super().__init__()
+
+        # Generic loss parameters.
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.data_range = data_range
+        
+        # Values from the paper
+        if scale_weights is None:
+            self.scale_weights = torch.tensor([0.096, 0.596, 0.289, 0.019])
+        else:
+            self.scale_weights = torch.tensor(scale_weights)
+
+        self.chromatic = chromatic
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+            
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=None)
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+            
+        scale_weights = self.scale_weights.to(prediction)
+        ms_gmds = []
+        for scale in range(4):
+            if scale > 0:
+                # Average by 2x2 filter and downsample
+                padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+                prediction = F.avg_pool2d(prediction, kernel_size=2, padding=padding)
+                target = F.avg_pool2d(target, kernel_size=2, padding=padding)
+        
+            score = _gmsd(prediction, target, reduction=None)
+            ms_gmds.append(score)
+        
+        # Concat results in different scales and multiply by weight
+        ms_gmds_val = scale_weights.view(1, 4, 1, 1) * (torch.cat(ms_gmds, dim=1) ** 2)
+        # Sum and take sqrt per-image
+        ms_gmds_val = torch.sqrt(torch.sum(ms_gmds_val, dim=-1))
```

Comment:
The sum is performed for `dim=4`, which equals 1. Shouldn't the operation be performed for `dim=(1,2,3)`?
```suggestion
        ms_gmds_val = torch.sqrt(torch.sum(ms_gmds_val, dim=(1, 2, 3)))
```